crates/library/std/src/evm.fe

```diff
@@ -0,0 +1,293 @@
+  unsafe:
+    return __balance(u256(addr))
+
+pub fn balance() -> u256:
+  unsafe:
+    return __selfbalance()
+
+pub fn caller() -> address:
+  unsafe:
+    return address(__caller())
+
+pub fn call_value() -> u256:
+  unsafe:
+    return __callvalue()
+
+
+# Overflowing math ops. Should these be unsafe or named
+# `overflowing_add`, etc?
+pub fn add(x: u256, y: u256) -> u256:
+  unsafe:
+    return __add(x, y)
+
+pub fn sub(x: u256, y: u256) -> u256:
+  unsafe:
+    return __sub(x, y)
+
+pub fn mul(x: u256, y: u256) -> u256:
+  unsafe:
+    return __mul(x, y)
+
+pub fn div(x: u256, y: u256) -> u256:
+  unsafe:
+    return __div(x, y)
+
+pub fn sdiv(x: u256, y: u256) -> u256:
+  unsafe:
+    return __sdiv(x, y)
+
+pub fn mod(x: u256, y: u256) -> u256:
+  unsafe:
+    return __mod(x, y)
+
+pub fn smod(x: u256, y: u256) -> u256:
+  unsafe:
+    return __smod(x, y)
+
+pub fn exp(x: u256, y: u256) -> u256:
+  unsafe:
+    return __exp(x, y)
+
+pub fn addmod(x: u256, y: u256, m: u256) -> u256:
+  unsafe:
+    return __addmod(x, y, m)
+
+pub fn mulmod(x: u256, y: u256, m: u256) -> u256:
+  unsafe:
+    return __mulmod(x, y, m)
+
+pub fn sign_extend(i: u256, x: u256) -> u256:
+  unsafe:
+    return __signextend(i, x)
+
+
+
+# Comparison ops
+# TODO: return bool (see issue #653)
+pub fn lt(x: u256, y: u256) -> u256:
+  unsafe:
+    return __lt(x, y)
+
+pub fn gt(x: u256, y: u256) -> u256:
+  unsafe:
+    return __gt(x, y)
+
+pub fn slt(x: u256, y: u256) -> u256:
+  unsafe:
+    return __slt(x, y)
+
+pub fn sgt(x: u256, y: u256) -> u256:
+  unsafe:
+    return __sgt(x, y)
+
+pub fn eq(x: u256, y: u256) -> u256:
+  unsafe:
+    return __eq(x, y)
+
+pub fn is_zero(x: u256) -> u256:
+  unsafe:
+    return __iszero(x)
+
+
+# Bitwise ops
+pub fn bitwise_and(x: u256, y: u256) -> u256:
+  unsafe:
+    return __and(x, y)
+
+pub fn bitwise_or(x: u256, y: u256) -> u256:
+  unsafe:
+    return __or(x, y)
+
+pub fn bitwise_not(x: u256) -> u256:
+  unsafe:
+    return __not(x)
+
+pub fn xor(x: u256, y: u256) -> u256:
+  unsafe:
+    return __xor(x, y)
+
+pub fn byte(offset: u256, x: u256) -> u256:
+  unsafe:
+    return __byte(offset, x)
+
+pub fn shl(x: u256, y: u256) -> u256:
+  unsafe:
+    return __shl(x, y)
+
+pub fn shr(x: u256, y: u256) -> u256:
+  unsafe:
+    return __shr(x, y)
+
+pub fn sar(x: u256, y: u256) -> u256:
+  unsafe:
+    return __sar(x, y)
+
+
+# Evm state access and control
+pub unsafe fn return_mem(location: u256, len: u256):
+    return __return(location, len)
+
+pub unsafe fn revert_mem(location: u256, len: u256):
+    return __revert(location, len)
+
+pub unsafe fn self_destruct(addr: address):
+    return __selfdestruct(u256(addr))
+
+# Invalid opcode. Equivalent to revert(0, 0),
+# except that all remaining gas in the current context
+# is consumed.
+pub unsafe fn invalid():
+    return __invalid()
+
+pub unsafe fn stop():
+    return __stop()
+
+pub unsafe fn pc() -> u256:
+    return __pc()
+
+# TODO: dunno if we should enable this
```

Comment:
What's the counter argument to not enable it?

Reply:
I'm not sure that I understand the semantics of yul's `pop`. It seems to just ignore the return value of an expression. Fe (and rust, without a `must_use` annotation) will happily ignore a return value that you don't use, but in yul that's an error. So fe's `call(...)` becomes `pop(call(...))` in yul.

Otherwise, it doesn't seem to do anything. Eg.
```
        let x := 5
        pop(x)
        mstore(0, x)
        return(0, 32)
```
becomes:
```
  0x05
  0x00
  mstore
  0x20
  0x00
  return
```
The `pop(x)` is apparently just 'ignoring the "return value" of the expression `x`', and thus a no-op.

So as far as I can tell, there's no point in including it.

If yul exposed raw pushN/pop/dupN etc stack manipulation operations, I'd certainly be in favor of including those in std::evm, but alas. It's a bummer that there's apparently no way we can allow advanced unsafe fe code to mess around with the stack.